### ISSUE TITLE
Limit / offset inner table with JOIN-ed ourter table

### ISSUE THREAD

**Issue by NoNameProvided** [typeorm/typeorm#711]
Hi @pleerock! 

## tl;dr; version

Is there a way to apply LIMIT and OFFSET on a sub-query via TypeORM? Something like 

```sql
SELECT * FROM (SELECT * FROM "article" "article" ORDER BY "article"."id" DESC LIMIT 2 OFFSET 0) as "distinct_article" LEFT JOIN "article_categories_category" "article_categories" ON "article_categories"."articleId"="distinct_article"."id" 
```

## Long version

I am looking for a way to apply `LIMIT` and `OFFSET` on inner table in a query with outer `INNER JOIN` or `LEFT JOIN`. My issue now is that `LIMIT` and `OFFSET` is applied on the full query. This can have an interesting side effect when having `@ManyToMany()` relationship.

So let's say I have an `Article` with a `ManyToMany` relation with `Categories` like this:

```
                                          Table "public.article"
        Column        |            Type             |                      Modifiers                       
----------------------+-----------------------------+------------------------------------------------------
 id                   | integer                     | not null default nextval('article_id_seq'::regclass)
 title                | character varying           | not null

                                     Table "public.category"
  Column   |            Type             |                       Modifiers                       
-----------+-----------------------------+-------------------------------------------------------
 id        | integer                     | not null default nextval('category_id_seq'::regclass)
 name      | character varying(20)       | not null


Table "public.article_categories_category"
   Column   |  Type   | Modifiers 
------------+---------+-----------
 articleId  | integer | not null
 categoryId | integer | not null

```

My content is: 
```sql
xxx=# TABLE article;
 id |   title         
----+------------
 12 | Article 12 
 13 | Article 13 
 14 | Article 14 
 15 | Article 15


xxx=# TABLE category;
 id |    name    
----+------------
  1 | Category 1
  2 | Category 2
  3 | Category 3


xxx=# TABLE article_categories_category;
 articleId | categoryId 
-----------+------------
        12 |          1
        12 |          2
        12 |          3
```

I can use queryBuilder to get my articles with limit and offset.

```ts
this.repository.createQueryBuilder('article')
   .offset(0).limit(2)
  .leftJoinAndSelect('article.categories', 'categories')
  .getMany();
```

However this will not return the value I would expect. The problem is this will return only one entity with the id of `12`. The reason is for that because TypeORM generates a query similar to this 

```sql
SELECT * FROM "article" "article" LEFT JOIN "article_categories_category" "article_categories" ON "article_categories"."articleId"="article"."id"  ORDER BY "article"."id" DESC LIMIT 2 OFFSET 0
```
(not exactly this, but this is the important part)

So because of the join and because there is two category for the article with `id = 12` it will return two row of `id = 12`. So when I apply `?limit=2&offset=0` TypeORM will be smart enough to merge the two object into one by it's id, **but** we get an different response as we would expect, while our entity has 3 category, our response will show it has only it has two: 

```json
{
     "id": 12,
     "title": "Article 12",
     "categories": [
     {
       "id": 2,
       "name": "Category 2"
     },
     {
       "id": 1,
       "name": "Category 1"
    }
  ]
}
```

Why? Because LIMIT is applied on the entire query while it should be applied on the inner SELECT.TypeORM should generate something like: 

```sql
SELECT * FROM (SELECT * FROM "article" "article" ORDER BY "article"."id" DESC LIMIT 2 OFFSET 0) as "distinct_article" LEFT JOIN "article_categories_category" "article_categories" ON "article_categories"."articleId"="distinct_article"."id" 
```

I dont know if this can be implemented, but it would be nice to have because there otherway as far as I understand LIMIT is not usable with the queryBuilder when using joins.

This report became a little bit long, I hope I could explain myself properly, if not, feel free to ask for clarification.

**Comment by pleerock** [comment]
Did you try [subquery functionality](https://github.com/pleerock/typeorm/blob/03ceaeade0d93c0884d1a85910e1bdc3ae461761/test/functional/query-builder/subquery/query-builder-subquery.ts#L19-L19) orm currently supports?

**Comment by NoNameProvided** [comment]
Haven't know about it, I will try it.  I will reopen if I am still facing issues.

Thanks!

**Comment by shvedovvladimir** [comment]
with subquery don't work getMany.

**Comment by bilelkh** [comment]
same problem :( for me

**Comment by DrEhsan** [comment]
For me when the relation between base entity and the relation entity is `OneToMany` this happens and `limit = 2` will result 2 records of relation entity. in this situation the `take` and `skip` will work correct.

But when the relation between base entity and the relation entity is `ManyToOne` the `limit` and `offset` are working as expected.
